### PR TITLE
Resource CEN

### PR DESCRIPTION
**Pull request by chumo-yl**
Resource CEN, can be used to create, destroy, modify and import a CEN resource. The results of acceptance tests are as below:

$TF_ACC=1 go test ./alicloud -v -run=TestAccAlicloudCen -timeout=300m
=== RUN   TestAccAlicloudCen_importBasic
--- PASS: TestAccAlicloudCen_importBasic (6.43s)
=== RUN   TestAccAlicloudCen_basic
--- PASS: TestAccAlicloudCen_basic (6.32s)
=== RUN   TestAccAlicloudCen_update
--- PASS: TestAccAlicloudCen_update (10.70s)
=== RUN   TestAccAlicloudCen_multi
--- PASS: TestAccAlicloudCen_multi (8.76s)
PASS
ok  	github.com/terraform-providers/terraform-provider-alicloud/alicloud	32.280s

### CODE REVIEW

File: alicloud/provider.go [hunk 1]
```diff
@@ -175,6 +175,7 @@ func Provider() terraform.ResourceProvider {
 			"alicloud_vpn_gateway":                 resourceAliyunVpnGateway(),
 			"alicloud_vpn_customer_gateway":        resourceAliyunVpnCustomerGateway(),
 			"alicloud_vpn_connection":              resourceAliyunVpnConnection(),
+			"alicloud_cen":                         resourceAlicloudCen(),
```
Comment (xiaozhu36) [inline alicloud/provider.go]: alicloud_cen 改为alicloud_cen_instance是不是会更好些

File: alicloud/resource_alicloud_cen.go [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+	var cen *cbn.CreateCenResponse
+	err := resource.Retry(3*time.Minute, func() *resource.RetryError {
+		args, err := buildAliCloudCenArgs(d, meta)
+		if err != nil {
+			return resource.NonRetryableError(fmt.Errorf("Building CreateVpcRequest got an error: %#v", err))
+		}
+		resp, err := client.cenconn.CreateCen(args)
+		if err != nil {
+			if IsExceptedError(err, CenQuotaExceeded) {
+				return resource.NonRetryableError(fmt.Errorf("The number of CEN exceeds the maximum value that can be created."))
```
Comment (xiaozhu36) [inline alicloud/resource_alicloud_cen.go]: 错误中把具体的错误详情也打印出来

File: alicloud/resource_alicloud_cen.go [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+	})
+	if err != nil {
+		return fmt.Errorf("Create cen got an error :%#v", err)
+	}
+
+	d.SetId(cen.CenId)
+
+	err = client.WaitForCen(d.Id(), Active, 60)
+	if err != nil {
+		return fmt.Errorf("Timeout when WaitForCenAvailable")
```
Comment (xiaozhu36) [inline alicloud/resource_alicloud_cen.go]: 92行把具体错误跑出来，不一定都是timeout

File: alicloud/resource_alicloud_cen.go [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+	}
+
+	return resourceAlicloudCenUpdate(d, meta)
```
Comment (xiaozhu36) [inline alicloud/resource_alicloud_cen.go]: 只有两个参数，并且这两个参数都是create的参数，所以create可以支持返回read

File: alicloud/resource_alicloud_cen.go [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+
+	if attributeUpdate {
+		if _, err := meta.(*AliyunClient).cenconn.ModifyCenAttribute(request); err != nil {
+			return err
+		}
+	}
+
+	d.Partial(false)
+
+	return resourceAlicloudCenRead(d, meta)
```
Comment (xiaozhu36) [inline alicloud/resource_alicloud_cen.go]: 如果update中只有一个api操作，那么可以把partial的设置都去掉

File: alicloud/resource_alicloud_cen.go [hunk 6]
```diff
@@ -0,0 +1,187 @@
...
+	request := cbn.CreateCreateCenRequest()
+
+	if v := d.Get("name").(string); v != "" {
+		request.Name = v
+	}
+
+	if v := d.Get("description").(string); v != "" {
+		request.Description = v
+	}
+
```
Comment (xiaozhu36) [inline alicloud/resource_alicloud_cen.go]: 把clienttoken加上，保证幂等性

File: alicloud/service_alicloud_cen.go [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+			log.Printf("Cen Id %s, Err %s", cenId, err)
+			return c, err
+		}
+
+		cenList := resp.Cens.Cen
+
+		for cenNum := 0; cenNum <= len(cenList)-1; cenNum++ {
+			if cenList[cenNum].CenId == cenId {
+				return cenList[cenNum], nil
+			}
```
Comment (xiaozhu36) [inline alicloud/service_alicloud_cen.go]: 可以直接将CenID作为API DescribeCens request的入参，省去循环判断

File: website/docs/r/cen.html.markdown [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+layout: "alicloud"
+page_title: "Alicloud: alicloud_cen"
+sidebar_current: "docs-alicloud-resource-cen"
+description: |-
+  Provides a Alicloud CEN resource.
+---
+
+# alicloud\_cen
+
+Provides a CEN resource.
```
Comment (xiaozhu36) [inline website/docs/r/cen.html.markdown]: CEN最好有全名解释，这个地方可加入https://www.alibabacloud.com/help/doc-detail/59870.htm作为refer

**Comment by chumo-yl** [comment]
modification finished. Test results:
$TF_ACC=1 go test ./alicloud -v -run=TestAccAlicloudCenInstance -timeout=300m
=== RUN   TestAccAlicloudCenInstance_importBasic
--- PASS: TestAccAlicloudCenInstance_importBasic (6.35s)
=== RUN   TestAccAlicloudCenInstance_basic
--- PASS: TestAccAlicloudCenInstance_basic (5.26s)
=== RUN   TestAccAlicloudCenInstance_update
--- PASS: TestAccAlicloudCenInstance_update (9.09s)
=== RUN   TestAccAlicloudCenInstance_multi
--- PASS: TestAccAlicloudCenInstance_multi (9.18s)
PASS
ok  	github.com/terraform-providers/terraform-provider-alicloud/alicloud	29.946s

File: alicloud/resource_alicloud_cen_instance.go [hunk 9]
```diff
@@ -0,0 +1,185 @@
...
+		args, err := buildAliCloudCenArgs(d, meta)
+		if err != nil {
+			return resource.NonRetryableError(fmt.Errorf("Building CreateVpcRequest got an error: %#v", err))
+		}
+		resp, err := client.cenconn.CreateCen(args)
+		if err != nil {
+			if IsExceptedError(err, CenQuotaExceeded) {
+				return resource.NonRetryableError(fmt.Errorf("Create CEN, the number of CEN instance exceeds the limit, got an error: %#v", err))
+			}
+			if IsExceptedError(err, OperationBlocking) || IsExceptedError(err, UnknownError) {
```
Comment (xiaozhu36) [inline alicloud/resource_alicloud_cen_instance.go]: Please use IsExceptedErrors to merge two errrors

File: alicloud/resource_alicloud_cen_instance.go [hunk 10]
```diff
@@ -0,0 +1,185 @@
...
+	})
+	if err != nil {
+		return fmt.Errorf("Create CEN and got an error: %#v", err)
+	}
+
+	d.SetId(cen.CenId)
+
+	err = client.WaitForCenInstance(d.Id(), Active, 60)
+	if err != nil {
+		return fmt.Errorf("WaitForCenAvailable and got an error: %#v", err)
```
Comment (xiaozhu36) [inline alicloud/resource_alicloud_cen_instance.go]: Please set resource id in the error message

File: alicloud/resource_alicloud_cen_instance.go [hunk 11]
```diff
@@ -0,0 +1,185 @@
...
+	return nil
+}
+
+func resourceAlicloudCenInstanceUpdate(d *schema.ResourceData, meta interface{}) error {
+	attributeUpdate := false
+	request := cbn.CreateModifyCenAttributeRequest()
+	request.CenId = d.Id()
+
+	if d.HasChange("name") {
+		d.SetPartial("name")
```
Comment (xiaozhu36) [inline alicloud/resource_alicloud_cen_instance.go]: please remove d.setpartial

File: alicloud/resource_alicloud_cen_instance.go [hunk 12]
```diff
@@ -0,0 +1,185 @@
...
+		request.Name = d.Get("name").(string)
+
+		attributeUpdate = true
+	}
+
+	if d.HasChange("description") {
+		d.SetPartial("description")
+		request.Description = d.Get("description").(string)
+
```
Comment (xiaozhu36) [inline alicloud/resource_alicloud_cen_instance.go]: there is no need blank line.

File: alicloud/resource_alicloud_cen_instance.go [hunk 13]
```diff
@@ -0,0 +1,185 @@
...
+		request.Name = v
+	}
+
+	if v := d.Get("description").(string); v != "" {
+		request.Description = v
+	}
+
+	request.ClientToken = buildClientToken("TF-CreateCenInstance")
+
+	return request, nil
```
Comment (xiaozhu36) [inline alicloud/resource_alicloud_cen_instance.go]: there is no need to return error.

**Pull request merged**